Sources/NIOConcurrencyHelpers/fastatomics.swift

```diff
@@ -0,0 +1,271 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import CNIOAtomics
+
+/// The protocol that all types that can be made atomic must conform to.
+///
+/// **Do not add conformance to this protocol for arbitrary types**. Only a small range
+/// of types have appropriate atomic operations supported by the CPU, and those types
+/// already have conformances implemented.
+public protocol FastAtomicPrimitive {
+    associatedtype AtomicWrapper
+    static var fast_atomic_create_with_existing_storage: (UnsafeMutableRawPointer, Self) -> Void { get }
```

Comment:
Is there any reason this can't take a typed pointer as its first argument?

Reply:
When I started, the ```AtomicWrapper``` type wasn't available on the Swift side, so I couldn't type it. But now that ```FutureAtomicPrimitive``` makes the type available, it shouldn't be a problem to change it.